spring-cloud-gateway-server-mvc/src/main/java/org/springframework/cloud/gateway/server/mvc/config/GatewayMvcPropertiesBeanDefinitionRegistrar.java

```diff
@@ -67,59 +38,31 @@
 import org.springframework.web.servlet.function.ServerResponse;
 
-import static org.springframework.cloud.gateway.server.mvc.handler.GatewayRouterFunctions.route;
-
 public class GatewayMvcPropertiesBeanDefinitionRegistrar implements ImportBeanDefinitionRegistrar {
 
-	private static final RequestPredicate neverPredicate = new RequestPredicate() {
-		@Override
-		public boolean test(ServerRequest request) {
-			return false;
-		}
-
-		@Override
-		public String toString() {
-			return "Never";
-		}
-	};
-
-	private static final RouterFunction<ServerResponse> NEVER_ROUTE = RouterFunctions.route(neverPredicate,
-			request -> ServerResponse.notFound().build());
-
-	protected final Log log = LogFactory.getLog(getClass());
-
-	private final TrueNullOperationArgumentResolver trueNullOperationArgumentResolver = new TrueNullOperationArgumentResolver();
-
-	private final Environment env;
-
-	private final FilterDiscoverer filterDiscoverer = new FilterDiscoverer();
-
-	private final HandlerDiscoverer handlerDiscoverer = new HandlerDiscoverer();
-
-	private final PredicateDiscoverer predicateDiscoverer = new PredicateDiscoverer();
-
-	private final ParameterValueMapper parameterValueMapper = new ConversionServiceParameterValueMapper();
-
-	public GatewayMvcPropertiesBeanDefinitionRegistrar(Environment env) {
-		this.env = env;
-	}
+	private final Log log = LogFactory.getLog(getClass());
 
 	@Override
 	public void registerBeanDefinitions(AnnotationMetadata importingClassMetadata, BeanDefinitionRegistry registry) {
 		// registers a RouterFunctionHolder that specifically isn't a RouterFunction since
 		// RouterFunctionMapping gets a list of RouterFunction and if you put
 		// RouterFunction in refresh scope, RouterFunctionMapping will end up with two.
-		// Uses this::routerFunctionHolderSupplier so when the bean is refreshed, that
-		// method is called again.
+		// Registers RouterFunctionHolderFactory::routerFunctionHolderSupplier so when the
+		// bean is refreshed, that method is called again.
 		AbstractBeanDefinition routerFnProviderBeanDefinition = BeanDefinitionBuilder
-				.genericBeanDefinition(RouterFunctionHolder.class, this::routerFunctionHolderSupplier)
+				.rootBeanDefinition(RouterFunctionHolder.class)
+				.setFactoryMethodOnBean("routerFunctionHolderSupplier", "routerFunctionHolderFactory")
 				.getBeanDefinition();
-		// TODO: opt out of refresh scope?
-		// Puts the RouterFunctionHolder in refresh scope
 		BeanDefinitionHolder holder = new BeanDefinitionHolder(routerFnProviderBeanDefinition,
 				"gatewayRouterFunctionHolder");
 		BeanDefinitionHolder proxy = ScopedProxyUtils.createScopedProxy(holder, registry, true);
-		routerFnProviderBeanDefinition.setScope("refresh");
+
+		// Puts the RouterFunctionHolder in refresh scope, if not disabled.
+		if (registry.containsBeanDefinition("refreshScope")) {
+			log.warn("Refresh scope is enabled. This is not supported in AOT mode. "
```

Comment:
We don't want a default application to have a warning. Please remove.

Reply:
Sure! Removed log.warn and (now) unused Logger and pushed